### PR TITLE
A4A-Partner Directory: Highlight invalid form fields

### PR DESCRIPTION
**Pull request by cleacos**
Resolves https://github.com/Automattic/automattic-for-agencies-dev/issues/716

## Proposed Changes

image

This PR highlights the invalid field values. In this case, we check/validate the URLs and the email.

The validator highlight text will be displayed if you type at least three (3) characters. 

**Note**

In a follow-up PR, the Agency Details form could be saved at any moment without completing the whole form, even with no valid fields.
- In another PR, on each field update, the details form will be saved.

## Why are these changes being made?

*

## Testing Instructions

- Check the code
- Try to introduce an invalid URL (Any form field for URL, in the Details and Expertise form)
- Try to introduce an invalid Email (Company email, in the Details form)

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by keoshi** [comment]
Looks good so far, @cleacos.

Have a few comments on the behavior and on styling though, so leaving it up to you whether to address here or on a separate PR.

### Behavior

The fields are marked as soon as it detects something is off, which is not a great experience. In other words, it points out the error even before a mistake happens. See this example where I'm trying to add a URL:

https://github.com/Automattic/wp-calypso/assets/390760/8dc86b0c-6609-42f2-9a50-58bb5ca75a88

Now check this example where it delays the error for when the input is de-selected (or `onblur`). Please note that this was the first example I found, but the particular implementation on this Codepen _does not matter_ for our case.

https://github.com/Automattic/wp-calypso/assets/390760/9df950ea-55ed-4a70-acbb-a3d9bcbdb8d4

From: https://codepen.io/bramus/pen/ExNYBOK

### Styling

In terms of styling, I think we can get rid of the parenthesis and also outline the entire field with the red color. That would look something like this:

image

Additionally, and if possible at all, we should try to position the error message below the field so that it's more obvious something needs attention. This is what we're doing [in Calypso](https://wpcalypso.wordpress.com/devdocs/design/form-fields):

image

Thanks, Clemente!

File: client/a8c-for-agencies/components/form/field/index.tsx [hunk 1]
```diff
@@ -8,6 +8,7 @@ type Props = {
 	sub?: string;
 	description?: string;
 	showOptionalLabel?: boolean;
+	validationText?: string;
```
Comment (jkguidaven) [inline client/a8c-for-agencies/components/form/field/index.tsx]: Based on the intention of the highlighting, I think maybe we can name this as `errorText` or `error` to emphasize that this is a negative effect.

**Comment by jkguidaven** [comment]
I completely agree with using the Calypso pattern to ensure consistency when displaying errors. We have already implemented this pattern on the 'Add Payment Method' page.

Screenshot 2024-06-26 at 11 44 48 PM

File: client/a8c-for-agencies/components/form/fieldset/index.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+type Props = {
+	label: string;
+	error?: string;
+	required?: boolean;
+	description?: string;
+	children?: ReactNode;
+};
+
+export default function A4AFormFieldset( {
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/components/form/fieldset/index.tsx:15]: This is created for the future, to have it aligned with all forms within A4A

**Comment by andrii-lysenko** [comment]
@keoshi @cleacos @jkguidaven  Can you give it a go. I've updated to follow signup pattern.

Screenshot 2024-06-26 at 12 44 32 PM

**Comment by cleacos** [comment]
Thank you @andrii-lysenko for working on it. I'll take a look and take over on it.

File: client/a8c-for-agencies/components/form/fieldset/index.tsx (hunk 2 above)
Comment (cleacos) [reply client/a8c-for-agencies/components/form/fieldset/index.tsx:15]: Why don't use it now?

File: client/a8c-for-agencies/sections/partner-directory/agency-details/hooks/use-details-form-validation.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import { useTranslate } from 'i18n-calypso';
+import { useCallback, useState } from 'react';
+import { AgencyDetails } from '../../types';
+
+export const CAPTURE_URL_RGX =
+	/^(http:\/\/www\.|https:\/\/www\.|http:\/\/|https:\/\/)?[a-z0-9]+([-.][a-z0-9]+)*\.[a-z]{2,63}(:[0-9]{1,5})?(\/.*)?$/i;
+
+const isValidUrl = ( urlString: string ) => {
+	return !! CAPTURE_URL_RGX.test( urlString );
+};
```
Comment (cleacos) [inline client/a8c-for-agencies/sections/partner-directory/agency-details/hooks/use-details-form-validation.ts]: We have the tools.ts file with the URL validator

File: client/a8c-for-agencies/sections/partner-directory/agency-details/index.tsx [hunk 4]
```diff
@@ -53,15 +56,32 @@ const AgencyDetailsForm = ( { initialFormData }: Props ) => {
...
+				if ( parent ) {
+					parent?.scrollTo( { behavior: 'smooth', top: 0 } );
+				}
+			}
+			return;
+		}
+
+		onSubmit();
+	};
+
```
Comment (cleacos) [inline client/a8c-for-agencies/sections/partner-directory/agency-details/index.tsx]: Maybe we could adopt Filipe's suggestion of validating when the input loses the focus (onBlur).

File: client/a8c-for-agencies/sections/partner-directory/agency-expertise/hooks/use-expertese-form-validation.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+type ValidationState = {
+	services?: string;
+	products?: string;
+	directories?: string;
+	clientSites?: string;
+	feedbackUrl?: string;
+};
+
+const useExperteseFormValidation = () => {
```
Comment (cleacos) [inline client/a8c-for-agencies/sections/partner-directory/agency-expertise/hooks/use-expertese-form-validation.ts]: `useExpertiseFormValidation` and the file name: `use-expertise-form-validation.ts`

File: client/a8c-for-agencies/sections/partner-directory/agency-details/index.tsx (hunk 4 above)
Comment (cleacos) [reply client/a8c-for-agencies/sections/partner-directory/agency-details/index.tsx]: Also, the agency details could be saved, even if they don't have all the fields filled out or are invalid.

**Review by keoshi — APPROVED** [review]
LGTM!

Would still try to move validation to `onblur` on each field instead of having to click the Submit button, but I understand it might be a bigger push for a later date.

**Comment by cleacos** [comment]
> Would still try to move validation to onblur on each field instead of having to click the Submit button

I was working on it, I prefer that solution too, onBlur... Since this PR is not super prioritize, I'll push this for tomorrow.

**Comment by cleacos** [comment]
After talking with @andrii-lysenko, we will address the right way to this, using the onBlur property and wrapping all this logic inside of the new `A4AFormFieldset` component.

With this version, the user cannot save the profile details until all the data is valid. The ideal behavior is that the user can save the profile information, but until we don't have the validation per each field instead of the full form, we don't have a way to highlight the wrong field. It's a short form, it's the MVP version, and highlighting what is the wrong field seems more important. We will address the right way in a follow-up PR. cc: @jeffgolenski

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16058124

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @cleacos for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16058124).

**Pull request merged**